[Feat]: Json Editor Component

Ayush8923 opened the pull request:
closes: #XX

### Visible/Frontend Changes

- [x] Init the Json Editor component

tomwwinter commented on src/app/core/common-components/json-editor/json-editor.component.ts:
```diff
@@ -0,0 +1,79 @@
+      props: {
+        content: { json: this.value || {} },
+        onChange: (updatedContent: any, { contentErrors }: any) => {
+          if (contentErrors?.length === 0) {
+            this.value = updatedContent.json;
+          }
```
What happen when contentErrors > 0? Should we display a notice here to the user?

Ayush8923 replied:
If `contentErrors > 0,` then something is wrong in JSON. I have made a few changes in the code to handle this error. Can you please take a look and let me know if any changes are needed?

Ayush8923 replied:
Thanks for this suggestion, it makes sense. I have updated it based on your suggestions. Now, the save button can only be pressed if the JSON is valid. Otherwise, the save button will remain disabled if the JSON is invalid.

**For reference:** I made some changes to the JSON to make it invalid, and the JSON editor component displayed an error indicating that the JSON is invalid and disabled the save button accordingly.

<img width="1254" alt="image" src="https://github.com/user-attachments/assets/401e5d36-6008-4ff0-b28e-41c63dfde252" />

sleidig replied:
:+1: 

You can remove the alert. The red error bar in the editor is very visible and should be sufficient.

Ayush8923 commented:
@sleidig Can you please review this PR and let me know if there are any changes needed.

sleidig commented on src/app/core/common-components/json-editor/json-editor.component.ts:
```diff
@@ -0,0 +1,71 @@
+   */
+  @Input() value: object = {};
+
+  @Input() height = "65vh";
```
can you design this component so that we can set any height through css on the `<app-json-editor style="height: 65vh"` and the json editor will fill whatever height it receives? That will make use much easier than having a custom input for the formatting here.

Ayush8923 replied:
Yes, I have updated the styling of the JSON editor component based on your suggestion. Now, by adding the height in the style from the parent component, we can manage the height of the JSON editor component directly from the parent. Please review the updated code and let me know if any changes are needed.

sleidig replied:
Great :slightly_smiling_face: 

The width of the popup jumps from narrow to wide and back (when errors are displayed it becomes wide). Can you fix an good explicit width for the MatDialog (maybe some "80vw"? if it works well on both desktop and mobile)

Ayush8923 replied:
I have fixed this, and added the width for the `MatDialog` which works well on both desktop and mobile. Can you please check and let me know if any further changes are needed?

sleidig replied:
works well now, thanks!

sleidig commented on src/app/features/notification/notification-settings/notification-settings.component.ts:
```diff
@@ -116,7 +116,9 @@ export class NotificationSettingsComponent implements OnInit {
 
   private async saveNotificationConfig(config: NotificationConfig) {
-    await this.entityMapper.save(config);
+    await this.entityMapper.save(config).catch((err) => {
+      Logging.error(err.message);
+    });
     this.alertService.addInfo($localize`Notification settings saved.`);
```
No!
This tells the user "settings saved" even if an error occured while saving, doesn't it?

Ayush8923 replied:
I have updated the code where the `Notification settings saved` alert was showing even if the data wasn't saved or if there was an error during the saving process. Now, if any error occurs while saving the data, it will display the error in the log.

sleidig reviewed (commented):
Seems to work fairly well, except when invalid json is entered (I added my thoughts to the comment thread for this). Can we open the "text" instead of the "tree" view by default?

I think ideally json-editor.component would extend `CustomFormControlDirective`, so that we can use this anywhere within a mat-form-field.
Not sure if implementing that will turn out too challenging for you @Ayush8923  but maybe you can give it a try.

Ayush8923 commented:
> Seems to work fairly well, except when invalid json is entered (I added my thoughts to the comment thread for this). Can we open the "text" instead of the "tree" view by default?
> 
> I think ideally json-editor.component would extend `CustomFormControlDirective`, so that we can use this anywhere within a mat-form-field. Not sure if implementing that will turn out too challenging for you @Ayush8923 but maybe you can give it a try.

Yes, by default, we can open the `text` tab instead of the `tree` tab. I have also updated this in the code. Please take a look and let me know if any changes are needed.

Sure, I am trying to extend the `CustomFormControlDirective` component into this JSON editor component.

Ayush8923 commented:
@sleidig I have extended the `CustomFormControlDirective` component into this JSON editor component. And also, I have made the changes according to your suggestions. Please take a look once and let me know if there are any other changes required. 

Additionally, I was trying to resolve the document conflict issue in this PR. The data is being updated correctly in the CouchDB document, and no conflicts are being shown in CouchDB. However, sometimes when saving data to CouchDB, a document conflict error is displayed.

sleidig commented on src/app/features/notification/notification-rule/notification-condition/notification-condition-editor/notification-condition-editor.component.html:
```diff
@@ -0,0 +1,26 @@
+    ></app-json-editor>
+
+    <div class="flex-row json-editor-actions">
+      <button mat-raised-button (click)="onJsonValueCancel()" i18n>
+        Cancel
+      </button>
+      <button
+        [disabled]="!isValidJson"
+        mat-raised-button
+        (click)="onJsonValueSave()"
+        i18n
+      >
+        Save
+      </button>
+    </div>
```
we generally have "Save" first, "Cancel" second in order

Ayush8923 replied:
I have updated it and made the changes accordingly.

sleidig commented on src/app/features/notification/notification-rule/notification-condition/notification-condition-editor/notification-condition-editor.component.spec.ts:
```diff
@@ -0,0 +1,61 @@
+    expect(component).toBeTruthy();
+  });
+
+  it("should close dialog when json is changed", () => {
```
really? This doesn't seem to happen (I have to click save, not only change the json - which is good this way, I think). Or do I misunderstand this test title?

sleidig commented on src/app/features/notification/notification-rule/notification-condition/notification-condition-editor/notification-condition-editor.component.ts:
```diff
@@ -0,0 +1,65 @@
+  templateUrl: "./notification-condition-editor.component.html",
+  styleUrl: "./notification-condition-editor.component.scss",
+})
+export class NotificationConditionEditorComponent {
```
Let's rename "JsonEditorDialogComponent" and move this to core. This can be reused for any kind of json editing

sleidig replied:
Can you extract this into its separate branch directly based into "master", also :blush: ?

Ayush8923 replied:
Sure, I have updated the component name according to your suggestion. 
Also, I have created the new [PR](https://github.com/Aam-Digital/ndb-core/pull/2798) for this which is based on `master`.

sleidig commented on src/app/features/notification/notification-rule/notification-rule.component.ts:
```diff
@@ -190,17 +191,57 @@ export class NotificationRuleComponent implements OnChanges {
+    }
+
+    const parsedObject = this.parseConditionsObjectToArray(result);
+    const conditionsFormArray = this.form.get("conditions") as FormArray;
+    conditionsFormArray.clear();
+    parsedObject.forEach((condition) => {
+      const conditionGroup = new FormGroup({
+        entityTypeField: new FormControl(condition.entityTypeField),
+        operator: new FormControl(condition.operator),
+        condition: new FormControl(condition.condition),
+      });
+      conditionsFormArray.push(conditionGroup);
+    });
```
shouldn't this be more or less the same as the initForm() method that we call onChanges?

Ayush8923 replied:
I have updated the code and reused the init form control. Please take a look on the updated changes and let me know if there are any further changes needed.

sleidig replied:
Thanks!

sleidig reviewed (approved):
JSON Editor seems to work well now :+1:
